pkg/util/checkpoint.go

```diff
@@ -0,0 +1,325 @@
+		Hash:   binary.BigEndian.Uint32(pkSha[:]),
+		Base64: base64.StdEncoding.EncodeToString(sig),
+	}
+
+	return &signature, nil
+
+}
+
+type SignedCheckpoint struct {
+	Checkpoint
+	// Signatures are one or more signature lines covering the payload
+	Signatures []note.Signature
+}
+
+// String returns the String representation of the SignedCheckpoint
+func (s SignedCheckpoint) String() string {
+	var b strings.Builder
+	b.WriteString(s.Checkpoint.String())
+	b.WriteRune('\n')
+	for _, sig := range s.Signatures {
+		var hbuf [4]byte
+		binary.BigEndian.PutUint32(hbuf[:], sig.Hash)
+		sigBytes, _ := base64.StdEncoding.DecodeString(sig.Base64)
+		b64 := base64.StdEncoding.EncodeToString(append(hbuf[:], sigBytes...))
+		fmt.Fprintf(&b, "%c %s %s\n", '\u2014', sig.Name, b64)
+	}
+
+	return b.String()
+}
+
+// UnmarshalText parses the common formatted checkpoint data and stores the result
+// in the SignedCheckpoint. THIS DOES NOT VERIFY SIGNATURES INSIDE THE CONTENT!
+//
+// The supplied data is expected to contain a single Checkpoint, followed by a single
+// line with no comment, followed by one or more lines with the following format:
+//
+// \u2014 name signature
+//
+// * name is the string associated with the signer
+// * signature is a base64 encoded string; the first 4 bytes of the decoded value is a
+//   hint to the public key; it is a big-endian encoded uint32 representing the first
+//   4 bytes of the SHA256 hash of the public key
+func (s *SignedCheckpoint) UnmarshalText(data []byte) error {
+	sc := SignedCheckpoint{}
+
+	if err := sc.Checkpoint.UnmarshalText(data); err != nil {
+		return errors.Wrap(err, "parsing checkpoint portion")
+	}
+
+	b := bufio.NewScanner(bytes.NewReader(data))
+	var pastCheckpoint bool
+	for b.Scan() {
+		if len(b.Text()) == 0 {
+			pastCheckpoint = true
+			continue
+		}
+		if pastCheckpoint {
+			var name, signature string
+			if _, err := fmt.Fscanf(strings.NewReader(b.Text()), "\u2014 %s %s\n", &name, &signature); err != nil {
+				return errors.Wrap(err, "parsing signature")
+			}
+
+			sigBytes, err := base64.StdEncoding.DecodeString(signature)
+			if err != nil {
+				return errors.Wrap(err, "decoding signature")
+			}
+			if len(sigBytes) < 5 {
+				return errors.New("signature is too small")
+			}
+
+			sig := note.Signature{
+				Name:   name,
+				Hash:   binary.BigEndian.Uint32(sigBytes[0:4]),
+				Base64: base64.StdEncoding.EncodeToString(sigBytes[4:]),
+			}
+			sc.Signatures = append(sc.Signatures, sig)
+		}
+	}
+	if len(sc.Signatures) == 0 {
+		return errors.New("no signatures found in input")
+	}
+
+	// copy sc to s
+	*s = sc
+	return nil
+}
+
+// Verify checks that one of the signatures can be successfully verified using
+// the supplied public key
+func (s SignedCheckpoint) Verify(public crypto.PublicKey) bool {
+	if len(s.Signatures) == 0 {
+		return false
+	}
+
+	msg, _ := s.Checkpoint.MarshalText()
+	//TODO: generalize this
+	digest := sha256.Sum256(msg)
+
+	var errorArray []error
```

Comment:
I don't think I follow the errorArray logic here. I'd expect the logic to be something like:

if all signatures verify, return true.
if any signature does not verify, return false.

The len(errorArray) == count check inside the loop has me confused a bit. It looks like it might return on the first successful verification.

Reply:
Yeah, the error handling can be simplified; There can be multiple signatures over the checkpoint content so if the public key verifies successfully against one of them then we return true.